Readme + script updates (#938)
Autodesk/synthesis: README.md

![Synthesis: An Autodesk Technology](/engine/Assets/Resources/Branding/Synthesis/Synthesis-An-Autodesk-Technology-2023-lockup-Blk-OL-No-Year-stacked.png#gh-light-mode-only)
![Synthesis: An Autodesk Technology](/engine/Assets/Resources/Branding/Synthesis/Synthesis-An-Autodesk-Technology-2023-lockup-Wht-OL-No-Year-stacked.png#gh-dark-mode-only)

<br/>

[![Engine](https://github.com/Autodesk/synthesis/actions/workflows/Engine.yml/badge.svg?branch=master)](https://github.com/Autodesk/synthesis/actions/workflows/Engine.yml)
[![API](https://github.com/Autodesk/synthesis/actions/workflows/API.yml/badge.svg?branch=master)](https://github.com/Autodesk/synthesis/actions/workflows/API.yml)
[![Clang Format](https://github.com/Autodesk/synthesis/actions/workflows/ClangFormat.yml/badge.svg?branch=master)](https://github.com/Autodesk/synthesis/actions/workflows/ClangFormat.yml)
[![Black Format](https://github.com/Autodesk/synthesis/actions/workflows/BlackFormat.yml/badge.svg?branch=master)](https://github.com/Autodesk/synthesis/actions/workflows/BlackFormat.yml)

> Synthesis is a robotics simulator designed by and for FIRST robotics students to help teams design, strategize, test and practice. Teams have the ability to import their own robots and fields using our [Fusion 360 Exporter](/exporter/) or use the pre-packaged ones that come included with every release of Synthesis.

Reviewer: ```suggestion
Synthesis is a robotics simulator designed by and for FIRST robotics students to help teams design, strategize, test and practice. Teams have the ability to import their own robots and fields using our [Fusion 360 Exporter](/exporter/SynthesisFusionAddin/) or use the pre-packaged ones that come included with every release of Synthesis.
```

> For more information on the product itself or the team, visit [http://synthesis.autodesk.com](http://synthesis.autodesk.com/).

Reviewer: ```suggestion
For more information on the product itself or the team, visit [https://synthesis.autodesk.com](https://synthesis.autodesk.com/).
```

- **Testing Robot Designs**: Synthesis is designed to be a tool for teams to quickly test their robot designs in a semi-realistic environment. Are you a builder who wants to use some crazy virtual four-bar linkage and your team says it's a waste of time? Well now you can prove them wrong by testing it in Synthesis!
- **Exploring the Field Environment**: Every year on kickoff, for both FTC and FRC FIRST competitions, Synthesis has a new release with the brand new field for that year included. This allows teams to explore the field through a 3D model, drive a robot around, and begin to strategize for the upcoming season's game.
- **Driver Practice & Strategy**: Not getting enough driver practice or don't have a full field available to you? Synthesis has you covered with the ability to drive your robot around with a gamepad from a first-person view at the driver station; allowing you to get a feel for potential control scheme layouts and any line-of-sight challenges that may arise. This also allows the drive team and the programmers to communicate about what control layouts work best for each driver.

## Getting Started

> If you are a FIRST robotics student who just wants to use Synthesis, you *don't* need this repo. Simply **install the latest release of Synthesis from [synthesis.autodesk.com/download](https://synthesis.autodesk.com/download.html)**.

Reviewer: ```suggestion
If you are a FIRST robotics student who just wants to use Synthesis, you *don't* need this repo. Simply **install the latest release of Synthesis from [synthesis.autodesk.com/download](https://synthesis.autodesk.com/download.html)** or from our [releases page](https://github.com/Autodesk/synthesis/releases).
```

If you're a developer who wants to contribute to Synthesis, you're in the right place. Synthesis is comprised of 3 main components that can be developed separately. These components include:

- [Synthesis API](/api/)
- [Simulation Engine](/engine/)

> - [Fusion Robot Exporter (Fusion 360 Plugin)](/exporter/)

Reviewer: ```suggestion
- [Fusion Robot Exporter (Fusion 360 Plugin)](/exporter/SynthesisFusionAddin/)
```

> Each of this components can be manually compiled separately, but this is not recommended. Instead, we recommend using the *init* scripts provided (`init.bat` & `init.sh` respectively) to build and link each component together (excluding the Fusion Robot Exporter).

Reviewer: ```suggestion
Each of this components can be manually compiled separately, but this is not recommended. Instead, we recommend using the *init* scripts provided (`init.bat` & `init.sh`) to build and link each component together (excluding the Fusion Robot Exporter).
```

### Compatibility Notes

As Fusion 360 is not supported on linux the linux installer does not come with the Fusion 360 Addin for exporting robots and fields.

Also note that the linux installer is tailored for Debian based distributions. Synthesis can be installed and ran on other distros but the installer is not guaranteed for your operating system.

### Dependencies

> Synthesis Version `6.0.0` has the following dependencies:

Reviewer: ```suggestion
Synthesis has the following dependencies:
```

This is just whatever is included in the repo, regardless of version

- [.NET Standard 2.0](https://learn.microsoft.com/en-us/dotnet/standard/net-standard?tabs=net-standard-1-0)
- [.NET 7.0](https://dotnet.microsoft.com/en-us/download/dotnet/7.0)
- [Protobuf 23.3](https://github.com/protocolbuffers/protobuf/releases/tag/v23.3)
- Unity Version 2022.3.2f1 (d74737c6db50)

> These dependencies need to be satisfied before attempting to build Synthesis with unity. You can either install these dependencies manually or use the *resolve dependencies* scripts (`scripts/win/resolve_deps.bat` & `scripts/osx/resolve_deps.sh` respectively) to install some of them for you.

Reviewer: ```suggestion
These dependencies need to be satisfied before attempting to build Synthesis with unity. You can either install these dependencies manually or use the *resolve dependencies* scripts (`scripts/win/resolve_deps.bat` & `scripts/osx/resolve_deps.sh`) to install some of them for you.
```

To automatically install .NET and Protobuf run the following script depending on your operating system:
- Windows: `scripts/win/resolve_deps.bat`
- MacOS & Linux: `scripts/osx/resolve_deps.sh`

Dependency Notes:

- The windows script requires admin privileges to install .NET and Protobuf to the system PATH. For this reason it is not included by default in `init.bat` unlike how `init.sh` includes `resolve_deps.sh`.
- Unity must be installed manually through the Unity Hub app which can be downloaded from [https://unity.com/download](https://unity.com/download).

### How To Build Synthesis Using An `init` Script And Unity

Before attempting to build Synthesis, ensure you have all dependencies installed.

1. Open a command prompt.
2. Change directories to a location where you'd like to clone the Synthesis repository.
3. Clone the Synthesis repository using `git clone https://github.com/Autodesk/synthesis --recurse-submodules`.
4. Change directories into the root of the Synthesis repository and run the `init` script for your operating system.
    - Windows: `init.bat`
    - MacOS & Linux: `init.sh`

> 5. Open `synthesis/engine` in Unity.

Reviewer: Maybe remove this and reference the engine setup?

Author: I like having it set up this way just so that everything is in one place and you don't need to hop around. But if you feel that it would be better we can.

Reviewer: Its fine, we can leave it

6. From there, you can run the simulation engine inside the Unity editor by opening `MainScene` from the `Assets/Scenes` directory or build it as a standalone application.
    - To build Synthesis as a standalone application, go to `File -> Build Settings` and select your target platform. Then, click `Build` and select a location to save the built application.

## Contributing

This project welcomes community suggestions and contributions. Synthesis is 100% open source and relies on the FIRST community to help make it better. The [Synthesis Contribution Guide](/CONTRIBUTING.md) suggests ways in which you can get involved through development and non-development avenues.

Before you contribute to this repository, please first discuss the change you wish to make via a GitHub issue, email us ([frc@autodesk.com](mailto:frc@autodesk.com)), or reach out through our [community discord](https://www.discord.gg/hHcF9AVgZA). This way we can ensure that there is no overlap between outside contributors and internal development work.

> When ready to contribute, just submit a pull request and be sure to include a clear and detailed description of the changes you've made so that we can verify them and eventually merge. Feel free to check out our [contributing guidelines](/CONTRIBUTING.md) to learn more.

Reviewer: ```suggestion
When ready to contribute, just submit a pull request and be sure to fill out the template accordingly so that we can verify them and eventually merge. Feel free to check out our [contributing guidelines](/CONTRIBUTING.md) to learn more.
```

## Code Formatting And Style

Synthesis uses [clang-format](https://clang.llvm.org/docs/ClangFormat.html) and [black](https://black.readthedocs.io/en/stable/) to format all of our `C++`, `C#` and `Python` code. Additionally we have GitHub workflows that verify these formatting standards for each and every pull request. For more information about how to run these formatters yourself, check out the [formatting tools](/scripts/format/) we use.

## Components

Here is a brief overview of each of the components found in this repository. For more information about each component, their dependencies and their individual build processes, check out their respective *READMEs*.

- [Synthesis API](/api/)
  - Components used to extend Synthesis' functionality while remaining Unity agnostic.
- [Code Emulation](/emulation/)
  - Synthesis tool designed to help users test their FRC robot code in a simulated environment. 
  - Code emulation is an advanced feature that is still under development, functionality and support is limited.
- [Core Engine](/engine/)
  - The core of Synthesis, the Unity project that contains the simulation engine.

> - [Fusion Exporter](/exporter/)

Reviewer: ```suggestion
- [Fusion Exporter](/exporter/SynthesisFusionAddin/)
```
